torch/_lobpcg.py

```diff
@@ -0,0 +1,671 @@
+            W = _utils.matmul(self.iK, self.R[:, nc:])
+
+            self.ivars['converged_end'] = ns = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+    def _update_ortho(self):
+        """
+        Update or initialize iteration variables when `method == "ortho"`.
+        """
+        mm = torch.matmul
+        ns = self.ivars['converged_end']
+        nc = self.ivars['converged_count']
+        n = self.iparams['n']
+        largest = self.bparams['largest']
+
+        if self.ivars['istep'] == 0:
+            Ri = self._get_rayleigh_ritz_transform(self.X)
+            M = _utils.qform(_utils.qform(self.A, self.X), Ri)
+            E, Z = _utils.symeig(M, largest)
+            self.X = mm(self.X, mm(Ri, Z))
+            self.update_residual()
+            np = 0
+            nc = self.update_converged_count()
+            self.S[:, :n] = self.X
+            W = self._get_ortho(self.R, self.X)
+            ns = self.ivars['converged_end'] = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+        else:
+            S_ = self.S[:, nc:ns]
+            # Rayleigh-Ritz procedure
+            E_, Z = _utils.symeig(_utils.qform(self.A, S_), largest)
+
+            # Update E, X, P
+            self.X[:, nc:] = mm(S_, Z[:, :n - nc])
+            self.E[nc:] = E_[:n - nc]
+            P = mm(S_, mm(Z[:, n - nc:], _utils.basis(_utils.transpose(Z[:n - nc, n - nc:]))))
+            np = P.shape[-1]
+
+            # check convergence
+            self.update_residual()
+            nc = self.update_converged_count()
+
+            # update S
+            self.S[:, :n] = self.X
+            self.S[:, n:n + np] = P
+            W = self._get_ortho(self.R[:, nc:], self.S[:, :n + np])
+            ns = self.ivars['converged_end'] = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+    def _get_rayleigh_ritz_transform(self, S):
+        """Return a transformation matrix that is used in Rayleigh-Ritz
+        procedure for reducing a general eigenvalue problem :math:`(S^TAS)
+        C = (S^TBS) C E` to a standard eigenvalue problem :math: `(Ri^T
+        S^TAS Ri) Z = Z E` where `C = Ri Z`.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        d = torch.zeros(d1.shape + (1, ),
+                        device=d1.device,
+                        dtype=d1.dtype)
+        d[:, 0] = d1
+        dd = mm(d, _utils.transpose(d))
+        d_ = mm(d, _utils.transpose(torch.ones_like(d)))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d_
```

Comment:
Confusing for me. Looks like `dd` is a matrix and its computation is quadratic in `n`. It is possible to eliminate this step and save some complexity by storing `BS` and computing
```
t1 = BS * d
t2 = S * d
R = torch.cholesky(t2^T * t1, upper=True)
```
This way we can save n^2 additive complexity factor, but will need a bit more memory though...

Reply:
Also, looks like `d_` contains copies of `d` as columns. Then `Rinv * d_` is a vector of all equal elements (can be eliminated to have additional n-factor). Is it really what it is supposed to do?

Reply:
Added a note on the RR implementation that should resolve the confusion.